### DISCUSSION TITLE
Is there any way for a non-logged in user to preview an unpublished page?

### DISCUSSION THREAD

**Discussion by mrpdean** [pimcore/pimcore#12162] [Q&A / Support] [answered]
Hi,

I'm wondering if there is any way to send a URL to someone which would allow them to view an unpublished page without having to log into the CMS, for the purpose of getting their approval to publish the page.

Thanks

**Comment by dvesh3** [comment]
Hi @mrpdean you can make a version public by clicking on public checkbox on versions tab:
![image](https://user-images.githubusercontent.com/5137917/169241685-ea129090-c8b5-4b39-9105-0dd5589cc123.png)

And the unauthenticated user then have to visit with page with specific version id like `pimcore.com/en/Magazine?v=versionId`

**Reply by mrpdean** [reply]
Hi @dvesh3 

Thanks for the info, unfortunately it doesn't seem to work.  It still throws an error when trying to access the version url from an unauthenticated browser session on my Pimcore instance.  I also tried it out on the Pimcore demo instance it also doesn't work: [https://demo.pimcore.fun/en/Magazine/Lifestyle?pimcore_version=3](https://demo.pimcore.fun/en/Magazine/Lifestyle?pimcore_version=3)

**Reply by dvesh3** [reply]
you have to use `?v=` param not `?pimcore_version=`. `pimcore_version` param requires admin session.

**Reply by mrpdean** [reply]
Yeah, I tried both.

?v= returns a 404
?pimcore_version= returns an access denied error

[https://10.x-dev.pimcore.fun/en/Magazine/Lifestyle?v=1](https://10.x-dev.pimcore.fun/en/Magazine/Lifestyle?v=1)

**Reply by dvesh3** [reply]
You are right, it only works when the document is published. Not sure why the publish status of a document is considered here.

**Reply by dvesh3** [reply]
So a little berdie told me that this feature was introduced keeping AB testing in mind and we don't want to expose the page which is not published yet. Sorry, for suggesting this feature in first place but it will not be useful for your case.

**Reply by mrpdean** [reply]
Thanks for the clarification @dvesh3.  

Curious about the AB testing feature but sadly cannot find any information about how to use it.  I don't suppose you have any info about it?

Anyway, I was able to work around preview limitation with a custom controller.  If anyone else stumbles across this, here is what I ended up doing.  Note that you might want to put some security around it, but I haven't done that yet.

```
<?php

namespace App\Controller;

use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;

use Pimcore\Controller\FrontendController;
use Pimcore\Config\Config;
use Pimcore\Model\Document;

class PreviewController extends FrontendController {

  /**
   * @Template()
   * @Route("/_preview/{id}", name="preview")
   */
  public function previewAction(Request $request, Config $websiteConfig, Int $id) {

    $document = Document::getById($id);

    if (!$document) {
      throw $this->createNotFoundException('Page not found');
    }

    return new Response(
      Document\Service::render($document)
    );
  }
}
```

**Comment by dvesh3** [comment, marked as answer]
> Anyway, I was able to work around preview limitation with a custom controller. If anyone else stumbles across this, here is what I ended up doing. Note that you might want to put some security around it, but I haven't done that yet.
> 
> ```
> <?php
> 
> namespace App\Controller;
> 
> use Symfony\Component\HttpFoundation\Request;
> use Symfony\Component\HttpFoundation\Response;
> use Symfony\Component\Routing\Annotation\Route;
> use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;
> 
> use Pimcore\Controller\FrontendController;
> use Pimcore\Config\Config;
> use Pimcore\Model\Document;
> 
> class PreviewController extends FrontendController {
> 
>   /**
>    * @Template()
>    * @Route("/_preview/{id}", name="preview")
>    */
>   public function previewAction(Request $request, Config $websiteConfig, Int $id) {
> 
>     $document = Document::getById($id);
> 
>     if (!$document) {
>       throw $this->createNotFoundException('Page not found');
>     }
> 
>     return new Response(
>       Document\Service::render($document)
>     );
>   }
> }
> ```

@mrpdean thanks for your answer!
